### PR TITLE
Phase change patch 2: variable coefficient and Marangoni surface tension

### PR DESCRIPTION
**Pull request by amneetb**
### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by amneetb** [comment]
@boyceg: Next stop patch 2. Should we rebase this branch against updated master or patch 1?

**Comment by boyceg** [comment]
> @boyceg: Next stop patch 2. Should we rebase this branch against updated master or patch 1?

I think that either would work, but I think that I would recommend rebasing against master at this point since the P1 branch has been merged into master.

I also think that you might find that it works better to set these PRs to target the previous PR rather than master, so that it is easier to see the changes that are specific to each PR. I always get nervous about changing the target branch for PRs though...

**Comment by drwells** [comment]
Lets just rebase against master at this point. That should be relatively straightforward - we just need to update the autotools files.

**Comment by Raamkrishnan** [comment]
I rebased this branch against the updated master. This patch is ready for review. The previous surface tension implementation could only handle a constant surface tension coefficient. Now, we modified it in such a way that it could handle both variable and constant surface tension coefficients. The variable surface tension coefficient is computed through a callback function. We have also added a new class ```MarangoniSurfaceTensionForceFunction``` derived from ```SurfaceTensionForceFunction``` class to compute the Marangoni forcing due to the surface tension gradients. Please take a look.

File: include/ibamr/SurfaceTensionForceFunction.h [hunk 1]
```diff
@@ -168,7 +168,53 @@ class SurfaceTensionForceFunction : public IBTK::CartGridFunction
...
+     * Get the ghost cell width of scratch data.
+     */
+    int getMinimumGhostWidth(const std::string& kernel_fcn);
+
+    const AdvDiffHierarchyIntegrator* const d_adv_diff_solver;
+    const SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_ls_var;
+    TimeSteppingType d_ts_type;
+    int d_C_idx, d_phi_idx;
+    std::string d_kernel_fcn;
+    double d_sigma = std::numeric_limits<double>::signaling_NaN(), d_num_interface_cells;
```
Comment (drwells) [inline include/ibamr/SurfaceTensionForceFunction.h:215]:
```suggestion
    double d_sigma = std::numeric_limits<double>::signaling_NaN(), d_num_interface_cells = std::numeric_limits<double>::signaling_NaN();
```

File: include/ibamr/SurfaceTensionForceFunction.h [hunk 2]
```diff
@@ -168,7 +168,53 @@ class SurfaceTensionForceFunction : public IBTK::CartGridFunction
...
+protected:
+    /*!
+     * Get the ghost cell width of scratch data.
+     */
+    int getMinimumGhostWidth(const std::string& kernel_fcn);
+
+    const AdvDiffHierarchyIntegrator* const d_adv_diff_solver;
+    const SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_ls_var;
+    TimeSteppingType d_ts_type;
+    int d_C_idx, d_phi_idx;
```
Comment (drwells) [inline include/ibamr/SurfaceTensionForceFunction.h:213]:
```suggestion
    int d_C_idx = IBTK::invalid_index, d_phi_idx = IBTK::invalid_index;
```

File: src/navier_stokes/fortran/navier_stokes_surface_tension_forcing2d.f.m4 [hunk 3]
```diff
@@ -281,17 +281,17 @@ ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
 c
 c     Compute surface tension forcing
 c
-c           F = sigma * K * grad(f)
+c           F = K * grad(f)
 c
+c     sigma will be multiplied later with F as F = sigma*F.
```
Comment (drwells) [inline src/navier_stokes/fortran/navier_stokes_surface_tension_forcing2d.f.m4:286]:
```suggestion
c     It is the caller's responsibility to multiply F by the (possibly
c     spatially-dependent) sigma coefficient
```

File: src/navier_stokes/fortran/navier_stokes_surface_tension_forcing2d.f.m4 [hunk 4]
```diff
@@ -319,32 +317,128 @@ c
...
-
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+c     Compute Marangoni force
+c
+c     F = grad T |grad C| - (grad T dot grad \phi)grad C
+c
+c     Marangoni coefficient will be multiplied later with F as
+c     F = marangoni_coef * F.
```
Comment (drwells) [inline src/navier_stokes/fortran/navier_stokes_surface_tension_forcing2d.f.m4:353]:
```suggestion
c     It is the caller's responsibility to multiply F by the (possibly
c     spatially-dependent) Marangoni coefficient.
```

**Review by drwells — CHANGES_REQUESTED** [review]
I have a few minor comments. There's a test that uses the new class in #1628 so lets fix these small things and then merge.

**Review by drwells — APPROVED** [review]

**Pull request merged**